### PR TITLE
Added build system check for posix_memalign support.

### PR DESCRIPTION
**Pull request by nmoinvaz**
Build system check I suggested in #884. I have added mentioned @concatime as a co-author since I took some of  the code from #884 for calling `posix_memalign`.

### CODE REVIEW

File: zutil_p.h [hunk 1]
```diff
@@ -5,15 +5,23 @@
 #ifndef ZUTIL_P_H
 #define ZUTIL_P_H
 
-#ifdef __APPLE__
+#if defined(HAVE_POSIX_MEMALIGN) && !defined(_POSIX_C_SOURCE)
+#  define _POSIX_C_SOURCE 200112L
```
Comment (nmoinvaz) [inline zutil_p.h]: I could have optionally set `_POSIX_C_SOURCE` in the build system, but I wanted to keep that localized to this file. I can always change it if needed.

**Comment by Dead2** [comment]
One of the CI checks errors out with
```
In file included from /home/runner/work/zlib-ng/zlib-ng/zutil.c:8:
/home/runner/work/zlib-ng/zlib-ng/zutil_p.h: In function ‘zng_alloc’:
/home/runner/work/zlib-ng/zlib-ng/zutil_p.h:25:12: error: implicit declaration of function ‘posix_memalign’ [-Werror=implicit-function-declaration]
   25 |     return posix_memalign(&ptr, 64, size) ? NULL : ptr;
      |            ^~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [CMakeFiles/zlib.dir/build.make:329: CMakeFiles/zlib.dir/zutil.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:360: CMakeFiles/zlib.dir/all] Error 2
make: *** [Makefile:160: all] Error 2
```

**Comment by mtl1979** [comment]
@Dead2 ```posix_memalign``` prototype is not defined if ```_POSIX_C_SOURCE``` is too low (as is the default)

**Comment by Dead2** [comment]
Needs a rebase

**Comment by nmoinvaz** [comment]
Rebased.

**Comment by nmoinvaz** [comment]
Not sure what include I am missing.

**Comment by mtl1979** [comment]
@nmoinvaz `_POSIX_C_SOURCE` might be defined to too low value.

File: zutil_p.h (hunk 1 above)
Comment (mtl1979) [reply zutil_p.h]: In some cases `_POSIX_C_SOURCE` is already set in build system, so this can fail if the previous value is too low.

**Comment by nmoinvaz** [comment]
@mtl1979 that doesn't appear to be the issue. I tried different values and also printed them out to make sure.

**Comment by nmoinvaz** [comment]
Merge #1019 first.

**Comment by mtl1979** [comment]
```
diff --git a/zutil.c b/zutil.c
index b38dc45a..602f946d 100644
--- a/zutil.c
+++ b/zutil.c
@@ -4,8 +4,8 @@
  */

 #include "zbuild.h"
-#include "zutil.h"
 #include "zutil_p.h"
+#include "zutil.h"

 z_const char * const PREFIX(z_errmsg)[10] = {
     (z_const char *)"need dictionary",     /* Z_NEED_DICT       2  */
```

**Comment by mtl1979** [comment]
Have to make sure no other file includes the header before defining `_POSIX_C_SOURCE`. ;)

**Comment by mtl1979** [comment]
Still failing CI...

**Comment by mtl1979** [comment]
Failing left and right...

**Comment by nmoinvaz** [comment]
Thanks dude I can see the CI for myself.

**Comment by mtl1979** [comment]
@nmoinvaz It's spamming other people with notifications...

**Comment by nmoinvaz** [comment]
There is an unsubscribe button on the right side.

File: zutil_p.h (hunk 1 above)
Comment (nmoinvaz) [reply zutil_p.h]: Fixed.

**Review by Dead2 — APPROVED** [review]
LGTM

**Pull request merged**